wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,431 @@
+tick ws = do
+    let wss               = ws  ^. wsState
+        currentSlot       = wss ^. wssCurrentSlot
+        rho               = _wsResubmissionFunction ws
+    case dueThisSlot currentSlot ws of
+         Left e -> throwM e
+         Right (dueNow, newSchedule) -> do
+            (toPrune, schedule') <- rho currentSlot dueNow newSchedule
+            let newState = ws & over (wsState . wssSchedule) (const schedule')
+                              . over (wsState . wssCurrentSlot) (mapSlot succ)
+            return (toPrune , newState)
+
+data SchedulingError =
+    LoopDetected Pending
+    -- ^ The transactions in this 'Pending' set forms a cycle and they
+    -- couldn't be top-sorted.
+
+instance Exception SchedulingError
+
+instance Show SchedulingError where
+    show (LoopDetected pending) = toString $ sformat ("LoopDetected " % F.build) pending
+
+-- | Returns a set of 'Pending' transactions which are due in the given
+-- 'Slot'.
+-- TODO(adn) Document this.
+--
+dueThisSlot :: Slot
+            -- ^ The current 'Slot'.
+            -> WalletSubmission m
+            -- ^ The 'WalletSubmissionState'.
+            -> Either SchedulingError ([ScheduledTx], Schedule)
+            -- ^ An error if no schedule can be produced, or the scheduled
+            -- transactions together with the new, updated 'Schedule'.
+dueThisSlot currentSlot ws =
+    let localPending = ws ^. wsState . wssPendingSet
+        (dependent, independent) = partitionScheduledTx localPending candidates
+        topSorted  = topsortTxs toTx independent
+    in case topSorted of
+            Nothing     -> Left (LoopDetected localPending)
+            Just sorted ->
+                let newSchedule = schedule { _ssUnsentNursery = dependent }
+                in Right (sorted, purgeSlot currentSlot newSchedule)
+    where
+        candidates :: [ScheduledTx]
+        candidates =
+            let due = fromMaybe mempty (IntMap.lookup (castSlot currentSlot) (_ssScheduled schedule))
+            in due `mappend` _ssUnsentNursery schedule
+
+        -- The current 'Schedule'.
+        schedule :: Schedule
+        schedule = ws ^. getSchedule
+
+        toTx :: ScheduledTx -> WithHash Core.Tx
+        toTx (ScheduledTx txId txAux _) =  WithHash (Core.taTx txAux) txId
+
+
+partitionScheduledTx :: Pending -> [ScheduledTx] -> ([ScheduledTx], [ScheduledTx])
+partitionScheduledTx pending xs = List.partition dependsOnPendingTx xs
+    where
+        -- | A 'ScheduledTx' is @not@ independent and should not be sent
+        -- over the wire if any of the inputs it consumes are mentioned in
+        -- the 'Pending' set.
+        dependsOnPendingTx :: ScheduledTx -> Bool
+        dependsOnPendingTx (ScheduledTx _ txAux _) =
+            let txs = pending ^. pendingTransactions . fromDb
+                inputs = List.foldl' updateFn mempty $ (Core.taTx txAux) ^. Core.txInputs . to NonEmpty.toList
+            in any (isJust . flip M.lookup txs) inputs
```

Comment:
Is this quite right? Transactions that depend on other pending transactions but are scheduled for the same slot are okay, no?

Reply:
great catch -- this is quite (too) aggressive; we should really be lenient and check, if we find a match, if the input is in the same slot, and if yes, we allow it to be sent.